File: Leetcode/mpsubarray.py
# leetcode no.152

class Solution:
    def maxProduct(self, nums) -> int:
        r_nums = list(reversed(nums))
        for i in range(1, len(nums)):
            nums[i] *= (nums[i-1] or 1)
            r_nums[i] *= (r_nums[i-1] or 1)
        return max(nums+r_nums)


s = Solution()
print(s.maxProduct([2, 3, -2, 4]))
print(s.maxProduct([-2, 0, -1]))

File: Leetcode/mpsubarray.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int size = nusm.size(), ret = nums[0], left = 0, right = 0;
        for(int i = 0; i < size; i++){
        	left = (left ? left : 1) * nums[i];
        	right = (right ? right : 1) * nums[n-i-1];
        	ret = max(ret, max(left, right));
		}
		return ret;
    }
};
